source/common/config/xds_mux/grpc_mux_impl.cc

```diff
@@ -0,0 +1,425 @@
+#include "source/common/config/xds_mux/grpc_mux_impl.h"
+
+#include "envoy/service/discovery/v3/discovery.pb.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/backoff_strategy.h"
+#include "source/common/config/decoded_resource_impl.h"
+#include "source/common/config/utility.h"
+#include "source/common/config/version_converter.h"
+#include "source/common/config/xds_context_params.h"
+#include "source/common/config/xds_resource.h"
+#include "source/common/memory/utils.h"
+#include "source/common/protobuf/protobuf.h"
+#include "source/common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Config {
+namespace XdsMux {
+
+namespace {
+class AllMuxesState {
+public:
+  void insert(ShutdownableMux* mux) {
+    absl::WriterMutexLock locker(&lock_);
```

Comment:
What is going on here? Shouldn't all muxes live on the main thread?

Reply:
I honestly didn't look very close, it's a verbatim copy of existing functionality (here: https://github.com/envoyproxy/envoy/blob/main/source/common/config/new_grpc_mux_impl.cc#L22). 

Said that, I agree, I don't think a mutex is needed there. @ggreenway: do you remember why a mutex is used here?

Reply:
I added it because I wasn't sure, was already dealing with a high-risk change, and wanted to make sure I didn't make anything worse. If you're sure it's not needed, I have no objection to removing it.

Reply:
@htuch: I'm going to remove the mutex in the unified mux implementation, and will do a seprate PR to remove it from the current muxes.

Reply:
Please also see https://github.com/envoyproxy/envoy/pull/17584.